sql/analyzer/resolve_subqueries.go

```diff
@@ -203,8 +203,18 @@ func isDeterminstic(n sql.Node) bool {
 
 // cacheSubqueryResults determines whether it's safe to cache the results for any subquery expressions, and marks the
 // subquery as cacheable if so. Caching subquery results is safe in the case that no outer scope columns are referenced,
-// and if all expressions in the subquery are deterministic.
+// if all expressions in the subquery are deterministic, and if the subquery isn't inside a trigger block.
 func cacheSubqueryResults(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
+	if n, ok := n.(*plan.TriggerBeginEndBlock); ok {
+		return plan.TransformExpressionsUpWithNode(n, func(n sql.Node, e sql.Expression) (sql.Expression, error) {
```

Comment:
No need to do a transform here

But you do need to look for TriggerBeginEndBlock anywhere in the tree, not just as a top-level node

Reply:
I think that since `cacheSubqueryResults` is called by a rule in the analyzer, it already traverses all the nodes of the tree. So it should find any TriggerBeginEndBlocks, and do the right thing. Also, I'm not sure if it's possible to have nested TriggerBeginEndBlocks.

Reply:
OK, now I understand why this works. Fun! You should write a comment indicating why this is valid (analyzer is recursively invoked on just the trigger body).

You should also write tests verifying this works (or doesn't) on AFTER triggers.